### PR TITLE
DRILL-1942-concurrency-test: new smoke test for concurrent query exec…

### PR DESCRIPTION
**Pull request by cwestin**
…ution;

useful to checking on the new allocator's locking schemes.

Made the SilentListener in BaseTestQuery.java public so it could be reused. Added new TestTpchDistributedConcurrent.java as a unit test. No other changes.

Unit tests (mvn install) pass.

### CODE REVIEW

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+  private final static int TOTAL_QUERIES = 115;
+  private final static int CONCURRENT_QUERIES = 15;
+
+  private final static Random random = new Random(0xdeadbeef);
+  private final static String alterSession = "alter session set `planner.slice_target` = 10";
+
+  private int remainingQueries = TOTAL_QUERIES - CONCURRENT_QUERIES;
+  private final Semaphore completionSemaphore = new Semaphore(0);
+  private final Semaphore submissionSemaphore = new Semaphore(0);
+  private final IdentityHashMap<UserResultsListener, Object> listeners = new IdentityHashMap<>();
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Guava's [Sets.newIdentityHashSet()](http://docs.guava-libraries.googlecode.com/git/javadoc/com/google/common/collect/Sets.html#newIdentityHashSet())?

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+
+    @Override
+    public void queryCompleted(QueryState state) {
+      super.queryCompleted(state);
+
+      final Object object;
+      synchronized(listeners) {
+        object = listeners.remove(this);
+      }
+      assertNotNull("listener not found", object);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: If this assert fails in the RPC thread, I don't think the test will fail. Can you confirm? IIRC I had similar issues in TestDrillbitResilience (see TestDrillbitResilience.WaitUntilCompleteListener#check).

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 1 above)
Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Is there any particular advantage to using that one instead? It sounds exactly the same.

Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: I don't know of any. `listeners.put(listener, listener)` kinda threw me off, and sets are nice constructs.

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 2 above)
Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: I tried it with a hard fail there to see what happens. The test does fail, but by timing out. Is that good enough?

Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Can you please add a comment next to the function call explaining this?

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 1 above)
Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: I had to make another change, so went ahead and did this as well.

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 2 above)
Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Done.

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+    @Override
+    public void queryCompleted(QueryState state) {
+      super.queryCompleted(state);
+
+      final boolean removed;
+      synchronized(listeners) {
+        removed = listeners.remove(this);
+      }
+
+      /*
```
Comment (jaltekruse) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Can we have some kind of error collector that we populate here and check upon submission of the next query? Making this timeout so long doesn't seem to be the right way to fix this problem.

Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Please help me out here: which timeout are you referring to? And where is "error collector that we populate here" referrring to? The pull request comment seems to refer to a block of code from a diff and not a single line.

Comment (jaltekruse) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Above there is this comment  "If the test does fail due to a timeout, see the comment in ChainingResultListener.queryCompleted() before assuming this needs to be adjusted."

I figured these were both referring to the same timeout, in which case I was trying to suggest we record the failure somewhere and check for it when we go to start the next query. Rather than throw an exception here, let the RPC system eat it and get stuck and let the test fail due to a timeout.

**Comment by cwestin** [comment]
I added code to collect errors, as well as to interrupt the test thread when an error occurs.

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java [hunk 4]
```diff
@@ -0,0 +1,206 @@
...
+    querySubmitter.start();
+
+    // Kick off the initial queries. As they complete, they will submit more.
+    submissionSemaphore.release(CONCURRENT_QUERIES);
+
+    // Wait for all the queries to complete.
+    try {
+      completionSemaphore.acquire(TOTAL_QUERIES);
+    } catch(InterruptedException e) {
+      fail("caught " + e);
```
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: If ```testThread``` is interrupted, this fail will trigger an ```AssertionError("...")```, however ```querySubmitter``` is not interrupted, and the ```System.err.println("...") ``` calls do not happen, right?

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 1 above)
Comment (sudheeshkatkam) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Did you revert this change?

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 4 above)
Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: I'll move the print to be above the fail(), and I'll move the fail out of the catch, but check a boolean from that instead. That will allow the querySubmitter to be interruped, and the errors will be printed. I'll wait a bit to see if there's anything else Jason wants as well.

File: exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java (hunk 1 above)
Comment (cwestin) [inline exec/java-exec/src/test/java/org/apache/drill/TestTpchDistributedConcurrent.java]: Not intentionally. I'll redo it when I make the change below around fail().

**Comment by cwestin** [comment]
Addressed Sudheesh's comments. Unless Jason has anything else, can we please get this merged now?

**Comment by parthchandra** [comment]
+1 looks like all comments are addressed.
